cmd/lp-cc/cc.go

```diff
@@ -0,0 +1,98 @@
+package main
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/ScottSallinen/lollipop/enforce"
+	"github.com/ScottSallinen/lollipop/graph"
+)
+
+const EMPTYVAL = math.MaxUint32
+
+type VertexProperty struct {
+	Value   uint32
+	Scratch uint32 // Intermediary accumulator
+}
+
+func (p *VertexProperty) String() string {
+	return fmt.Sprintf("%d", p.Value)
+}
+
+type EdgeProperty struct {
+	Weight float64
+}
+
+type MessageValue uint32
+
+func MessageAggregator(dst *graph.Vertex[VertexProperty, EdgeProperty], didx, sidx uint32, data MessageValue) (newInfo bool) {
```

Comment:
Got it.

Reply:
ScottSallinen7 hours ago
func MessageAggregator(dst *graph.Vertex[VertexProperty, EdgeProperty], didx, sidx uint32, data MessageValue) (newInfo bool) {
    input := uint32(data)
    if input > dst.Id { // This means ininitalization
        input = dst.Id
    }
Because value should not be greater than Id.

Reply:
Either way would be technically correct, but change the comment as appropriate.
1.
```
if input == EMPTYVAL { // This means initialization
		input = dst.Id
}
```
This one is correct, as it only targets the initialization message (g.InitVal) to change it to the vertex's own ID.

2.
```
if input > dst.Id { // We only care if a message is lower than our own ID
    input = dst.Id
}
```
The more generic way, does not specifically target the initialization message and will apply to any incoming message changing it.

I prefer 1, because it does not require a vertex dereference and does less work overall.

Reply:
if input == EMPTYVAL { // This means initialization
        input = dst.Id
}

This can fail in asynchronous situations, as the logs show.
As written above, a node's initial value will be EMPTYVAL, and it will be initialized to its own label only when it is detected that it is EMPTYVAL.
The problem is that if node y adjacent to node x updates x with its label, but y's label is greater than x, x will not initialize the label to its own, which can cause serious problems.
So I suggest that it should be written according to my method.

logs:

=== RUN   TestAsyncDynamic
2022/09/30 21:26:03 [Framework]	 ConvergeAsyncDynWithRate
2022/09/30 21:26:03 [Graph]	 Read 11 edges in (ms) 2 fromWatch 2.3741ms
2022/09/30 21:26:03 [Framework]	 T0EdgeFinish 2
2022/09/30 21:26:03 [Framework]	 Termination(ms) 2 realtime(ms): 2
2022/09/30 21:26:03 [CC]	 vertex values: 
2022/09/30 21:26:03 [CC]	 # vidx = 0 Id = 7 value = 0
2022/09/30 21:26:03 [CC]	 # vidx = 1 Id = 0 value = 0
2022/09/30 21:26:03 [CC]	 # vidx = 2 Id = 3 value = 0
2022/09/30 21:26:03 [CC]	 # vidx = 3 Id = 8 value = 0
2022/09/30 21:26:03 [CC]	 # vidx = 4 Id = 1 value = 2
2022/09/30 21:26:03 [CC]	 # vidx = 5 Id = 4 value = 2
2022/09/30 21:26:03 [CC]	 # vidx = 6 Id = 2 value = 2
2022/09/30 21:26:03 [CC]	 # vidx = 7 Id = 5 value = 2
2022/09/30 21:26:03 [CC]	 # vidx = 8 Id = 9 value = 0
2022/09/30 21:26:03 [CC]	 # vidx = 9 Id = 6 value = 2
    cc_test.go:31: 4  is  2  expected  1
    cc_test.go:31: 6  is  2  expected  1
    cc_test.go:31: 5  is  2  expected  1
    cc_test.go:31: 7  is  2  expected  1
    cc_test.go:31: 9  is  2  expected  1

Reply:
This difference was due to a minor difference in the flow of initialization with dynamic graphs, which was fixed in this commit:
https://github.com/ScottSallinen/lollipop/commit/f554abc27d01ced2d84c33aefd1030feb0d446ec
If you rebase to the latest master, you will not see this issue.

While you are correct that that X could Y's value before it is initialized, the initialization message (which is at most once but not guaranteed to be first), the initialization message when it arrives will indeed correct X to it's own value.

So you can do it as I suggest after rebasing to include the above commit.